src/app/CommandSender.cpp

```diff
@@ -61,10 +62,22 @@ CHIP_ERROR CommandSender::SendCommandRequest(NodeId aNodeId, Transport::AdminId
     VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
     mpExchangeCtx->SetResponseTimeout(kImMessageTimeoutMsec);
 
-    err = mpExchangeCtx->SendMessage(
-        Protocols::InteractionModel::MsgType::InvokeCommandRequest, std::move(mCommandMessageBuf),
-        Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse).Set(Messaging::SendMessageFlags::kNoAutoRequestAck));
-    SuccessOrExit(err);
+    event.Type                              = DeviceLayer::DeviceEventType::kInteractionModelCommandRequest;
+    event.InteractionModelEvent.ExchangeCtx = mpExchangeCtx;
+
+    // ChipDeviceEvent can only contains the structures with trivial copy assignment operator, so we can't use PacketBufferHandle
+    // and SendFlags directly within ChipDeviceEvent.
+    event.InteractionModelEvent.Payload   = std::move(mCommandMessageBuf).UnsafeRelease();
+    event.InteractionModelEvent.SendFlags = Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse)
+                                                .Set(Messaging::SendMessageFlags::kNoAutoRequestAck)
+                                                .Raw();
+
+    // Currently, Interaction Model protocol is implemented in app thread instead of CHIP thread, we need to lock CHIP stack to
+    // prevent it from operating before we post an event to the event queue associated with the CHIP thread.
+    DeviceLayer::PlatformMgr().LockChipStack();
+    DeviceLayer::PlatformMgr().PostEvent(&event);
```

Comment:
PostEvent is thread safe

Reply:
We also need to stop event processing in CHIP stack in this case.

Reply:
Why? The event loop has a queue. If the queue is protected (which it IS in FreeRTOS and Zephyr), it should be safe.

Right now the problem is that the CHIP event loop in POSIX is not safe, so you need that lock, but we should not need locks around posting to the one queue.

Reply:
Yes, we should fix the design pattern issue instead of continue to use WR. The CHIP event queue need to be protected with a dedicated mutex.